### ISSUE TITLE
Inferred function return type is different from TS 5.9.3

### ISSUE THREAD

**Issue by In3luki** [microsoft/typescript-go#2161]
## Steps to reproduce

This is a contrived reproduction from a larger codebase. It looks like the error happens on line 24 where the inferred type of `getItem(options.uuid)` is always the same type as `options.item` as far as I can tell.

https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDiCmA7eAnAlsAkgF3gW2gG8AoaM6RMPeaAXmgCIBzJVDBgbmIF9jjRIMAMoAHeMDQAzDDnzR4AD1yIAJjATJ0WXARLloAewBuqdCpqVqdRhDETpwTjz6SAromDY0BxNBbZZPAAeABUAEQNgV2pEbHklJDU4Vi1AgD4AClJyAH1XVzQVAC5oCGx0RCYAGmIAShLwyOikOIAfClcQECJsshR4bFcUX0IKKngS5hT2aG5oSGhGqJjsLl5iNw8vHz8B0XEpGR0M+uh9+yO5PXJ+weGiMepJ2wOHBlm1vmAfMsMRbcQEBK1zI+UKAH4SmUKkwuPo0DpIWc7IdtHJ2ohOiA1tZCNwuPwfnEEXJ6AZ-t5AQA6EkEcHg6AZckAiBUsEqaAM-yBJkUn5sgoqWrQEr+c6onm1WoEqSMgCEGVp0DQgOwYA88AMkmSmkueClPX02AAFigDAB3aAAURQZpQGQYADkDHF1cqdHKGNKePIQBAaLKMgqlSqyurgJrteKHIEDSDoN9AQYQPAqeawMNFToqZZ4N7eEA

## Behavior with `typescript@5.9`

The type of `item` is correctly narrowed to `GenericItem` at the end of the if block and no errors are shown.

## Behavior with `tsgo`

The type of `item` becomes `never` and the error `Property 'name' does not exist on type 'never'` is shown.

**Comment by Andarist** [comment]
This boils down to "Improved inference from return types (old code base is inconsistent here)." introduced by https://github.com/microsoft/typescript-go/pull/244
 
```ts
declare const obj: {
  prop: number | undefined;
};

declare function returnTypeInference<T>(): T;
const result1 = obj.prop ?? returnTypeInference();
//                          ^? Strada: function returnTypeInference<number | undefined>(): number | undefined
//                             Corsa: function returnTypeInference<number | undefined>(): number | undefined

declare function nonNullableReturnTypeInference<T extends {}>(): T;
const result2 = obj.prop ?? nonNullableReturnTypeInference();
//                         ^? Strada: function nonNullableReturnTypeInference<{}>(): {}
//                            Corsa: function nonNullableReturnTypeInference<number>(): number
```

There is an old PR of mine that implements basically the same thing in Strada [here](https://github.com/microsoft/TypeScript/pull/58910). By landing it, we could at least make versions behave the same way 😉

**Comment by jakebailey** [comment]
Going to close this now; https://github.com/microsoft/TypeScript/pull/58910 is merged and 6.0 will match.
